### PR TITLE
supp domain generation

### LINKED ISSUE #111: Feature Request:  Split to SUPP-- domain functionality
### Feature Idea

Create function to derive SUPP-- domain from the SDTM output

- Users to provide list of SUPP variables and their QORIG
- `pivot_longer` to shape the data frame
- Columns to include in the output:
      "STUDYID" - inherit from SDTM output
      "RDOMAIN" - inherit from DOMAIN variable
      "USUBJID" - inherit from SDTM output
      "IDVAR" - user input, default to SEQ
      "IDVARVAL" - values from IDVAR
      "QNAM" - user input, SUPP variable names
      "QVAL" - SUPP variable values
      "QORIG" - user input
      "QEVAL" - user input, default to NA

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

_No response_

### PR DESCRIPTION
**Pull request by ShiyuC**
Close #111 

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull z
Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the 
[tidyverse style guide](https://style.tidyverse.org/). Run 
`styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should
consider realistic data scenarios and edge cases, e.g. empty datasets, errors,
boundary cases etc. - See 
[Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you 
fully follow the 
[deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords
and families. Refer to the 
[categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the
`NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it
has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build oak site `pkgdown::build_site()` and check that all affected
examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

File: R/gen_supp.R [hunk 1]
```diff
@@ -0,0 +1,113 @@
+#' Function to generate supplemental domains SUPP--
+#'
+#' @param sdtm_dataset SDTM output used to split supplemental domains.
+#' @param idvar Variable name used to for variable.
+#' @param spec Data frame of specifications..
```
Comment (ShiyuC) [inline R/gen_supp.R:5]: Try to incorporate `xportr_spec`; Let the user define the input of label and origin

Comment (ShiyuC) [reply R/gen_supp.R]: Used the naming convention from xportr to create spec example

File: R/gen_supp.R [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+#' @examples
+#' path <- here::here("inst")
+#' dm <- readRDS(paste0(path, "/domain/dm.rds"))
+#' spec <- read.csv(paste0(path, "/spec/spec.csv"))
+#' suppdm <- gen_supp(dm, idvar = NULL, spec = spec, eval = "CLINICAL STUDY SPONSOR")
+gen_supp <-
+  function(sdtm_dataset = NULL,
+           idvar = NULL,
+           spec = NULL,
+           eval = NA_character_ ### Should this be 1 single character value or a vector?
```
[the remark this answers is not in the document]
Comment (ShiyuC) [reply R/gen_supp.R]: SME confirmed the QEVAL values can vary within one single supp domain

File: R/gen_sdtm_supp.R [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+#' `qnam_var`, `label_var` and `orig_var`.
+#' @param qnam_var Variable name in user-defined `spec` for QNAM variable.
+#' @param label_var Variable name in user-defined `spec` for QLABEL variable.
+#' @param orig_var Variable name in user-defined `spec` for QORIG variable.
+#'
+#' @return List of SDTM domain with suppquals dropped and corresponding supplemental domain.
+#' @export
+#'
+#' @examples
+#' path <- here::here("inst")
```
Comment (ShiyuC) [inline R/gen_sdtm_supp.R:15]: The here package won't work for the users. Switch to system.file

**Comment by rammprasad** [comment]
Looks good to me @ShiyuC - Can you merge main to FB?

File: R/gen_sdtm_supp.R [hunk 4]
```diff
@@ -0,0 +1,142 @@
+#' Function to generate final SDTM domain and supplemental domain SUPP--
+#'
+#' @param sdtm_dataset SDTM output used to split supplemental domains.
+#' @param idvar Variable name for IDVAR variable.
+#' @param spec User-defined data frame of specifications which should contain
+#' `qnam_var`, `label_var` and `orig_var`.
```
Comment (rammprasad) [inline R/gen_sdtm_supp.R:6]: Can we change the name of the parameter to `supp_qual_info`

**Review by rammprasad — CHANGES_REQUESTED** [review]
Minor comments.

**Comment by ShiyuC** [comment]
@edgar-manukyan Can I request your review of this PR? Thanks!

File: R/gen_sdtm_supp.R [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        values_transform = as.character
+      ) |>
+      # Exclude the records where QVAL is null
+      dplyr::filter(!is.na(QVAL)) |>
+      tidyr::pivot_longer(
+        cols = dplyr::all_of(idvar),
+        names_to = "IDVAR",
+        names_transform = function(names_values) {
+          # Replace DMSEQ with NA
+          names_values <- gsub("^DMSEQ$", NA_character_, names_values)
```
Comment (edgar-manukyan) [inline R/gen_sdtm_supp.R:83]:
```suggestion
          gsub("^DMSEQ$", NA_character_, names_values)
```

File: R/gen_sdtm_supp.R [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    for (v in seq_len(length(labels))) {
+      attr(supp[[v]], "label") <- labels[v]
+    }
+
+    sdtm_output <- sdtm_dataset |>
+      dplyr::select(-dplyr::all_of(supp_cols), -dplyr::any_of("DMSEQ"))
+
+    domain <- toupper(domain)
+    supp_domain <- paste0("SUPP", domain)
+    final <- rlang::list2(
```
Comment (edgar-manukyan) [inline R/gen_sdtm_supp.R:134]:
```suggestion

    final <- rlang::list2(
```

**Comment by edgar-manukyan** [comment]
> @edgar-manukyan Can I request your review of this PR? Thanks!

Nice job @ShiyuC 💯 💯 💯 I left few stylistic comments, suggestions.

File: R/gen_sdtm_supp.R [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+#' @param label_var Variable name in user-defined `supp_qual_info` for QLABEL variable.
+#' @param orig_var Variable name in user-defined `supp_qual_info` for QORIG variable.
+#'
+#' @return List of SDTM domain with suppquals dropped and corresponding supplemental domain.
+#' @export
+#'
+#' @examples
+#' dm <- read_domain_example("dm")
+#' supp_qual_info <- read.csv(system.file("spec/suppqual_spec.csv", package = "sdtm.oak"))
+#' final <-
```
[the remark this answers is not in the document]
Comment (rammprasad) [reply R/gen_sdtm_supp.R:17]: Can we rename the object to be something meaningful? `dm_suppdm`

**Review by rammprasad — CHANGES_REQUESTED** [review]

File: R/gen_sdtm_supp.R [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    # Create vectors for later use
+    domain <- unique(sdtm_dataset$DOMAIN)
+    assertthat::assert_that(identical(length(domain), 1L),
+      msg = "There are multiple domain names in the SDTM dataset"
+    )
+
+    # Add dummy SEQ variable for DM domain.
+    # The DMSEQ will be removed later.
+    if (identical(domain, "DM") && is.null(idvar)) {
+      sdtm_dataset$DMSEQ <- NA_integer_
```
[the remark this answers is not in the document]
Comment (ShiyuC) [reply R/gen_sdtm_supp.R:50]: From my understanding, the `names_transform` in `pivot_longer` doesn't change the values under `DMSEQ` to be character NA, it will just change the "DMSEQ" value itself into `NA` under the `IDVAL` column because there is no such SDTM variable called "DMSEQ".

File: R/gen_sdtm_supp.R [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+#'     orig_var = "Origin"
+#'   )
+gen_sdtm_supp <-
+  function(sdtm_dataset,
+           idvar = NULL,
+           supp_qual_info,
+           qnam_var,
+           label_var,
+           orig_var) {
+    admiraldev::assert_character_scalar(qnam_var)
```
[the remark this answers is not in the document]
Comment (ShiyuC) [reply R/gen_sdtm_supp.R:34]: Updated

**Comment by ShiyuC** [comment]
@edgar-manukyan @rammprasad the PR is updated. Please conduct another review when you get a chance. 😺

File: R/generate_sdtm_supp.R [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    ) |>
+    # Exclude the records where QVAL is null
+    dplyr::filter(!is.na(QVAL)) |>
+    tidyr::pivot_longer(
+      cols = dplyr::all_of(idvar),
+      names_to = "IDVAR",
+      names_transform = function(names_values) {
+        # Replace DMSEQ with NA
+        names_values <- gsub("^DMSEQ$", NA_character_, names_values)
+        names_values
```
Comment (edgar-manukyan) [inline R/generate_sdtm_supp.R:86]: I do not think we need to assign the result of gsub, it will just return itself.

**Review by edgar-manukyan — APPROVED** [review]
@ShiyuC, thanks for the updates. Approved. Just left two comments, more like questions for my own understanding. 🙏

File: R/generate_sdtm_supp.R [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+  domain <- unique(sdtm_dataset$DOMAIN)
+
+  assertthat::assert_that(identical(length(domain), 1L),
+    msg = "There are multiple domain names in the SDTM dataset"
+  )
+
+  # Add dummy SEQ variable for DM domain.
+  # We are removing DMSEQ later in the code.
+  if (identical(domain, "DM") && is.null(idvar)) {
+    sdtm_dataset$DMSEQ <- NA_integer_
```
[the remark this answers is not in the document]
Comment (ShiyuC) [reply R/generate_sdtm_supp.R:49]: @edgar-manukyan , the logic is inherited from roak but let me try to explain it better. For DM, this is what we are expecting to do:
image

IDVAR is a character variable while IDVARVAL is a numeric variable. Please let me know if this makes sense.

File: R/generate_sdtm_supp.R (hunk 10 above)
Comment (ShiyuC) [reply R/generate_sdtm_supp.R]: Removed the return

**Review by rammprasad — APPROVED** [review]
